### PR TITLE
Network

### PR DESCRIPTION
**Pull request by joakim-hove**
Added input support for a network model, based on the keywords BRANPROP and NODEPROP

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
opinions?

File: src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+void ExtNetwork::add_branch(Branch branch)
+{
+    this->m_branches.push_back( std::move(branch) );
+}
+
+
+std::optional<Branch> ExtNetwork::uptree_branch(const std::string& node) const {
+    if (!this->has_node(node))
+        throw std::out_of_range("No such node: " + node);
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp:78]: You should
```C++
#include <stdexcept>
```
to bring in a declaration of `std::out_of_range`.

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+public:
+    ExtNetwork() = default;
+    bool active() const;
+    bool has_node(const std::string& name) const;
+    void add_node(Node node);
+    void add_branch(Branch branch);
+    const Node& node(const std::string& name) const;
+    const Node& root() const;
+    std::vector<Branch> downtree_branches(const std::string& node) const;
+    std::optional<Branch> uptree_branch(const std::string& node) const;
```
Comment (bska) [inline opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:44]: You should
```C++
#include <optional>
```
to bring in a declaration of `std::optional`.

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Opm {
+namespace Network {
+
+
+class ExtNetwork {
+public:
+    ExtNetwork() = default;
+    bool active() const;
+    bool has_node(const std::string& name) const;
```
Comment (bska) [inline opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:38]: You should
```C++
#include <string>
```
to bring in a declaration of `std::string`.

File: opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef NETWORK_NODE_HPP
+#define NETWORK_NODE_HPP
+
+#include <optional>
```
Comment (bska) [inline opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp:24]: You should
```C++
#include <string>
```
to bring in a declaration of `std::string`.

File: src/opm/parser/eclipse/EclipseState/Schedule/Network/Node.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+Node::Node(const std::string& name, std::optional<double> terminal_pressure, bool upstream_choke, bool add_gas_lift_gas, const std::string& choke_target_group) :
+    m_name(name),
+    m_terminal_pressure(terminal_pressure),
+    m_upstream_choke(upstream_choke),
+    m_add_gas_lift_gas(add_gas_lift_gas),
+    m_choke_target_group(choke_target_group)
+{
+    if (!upstream_choke)
+        throw std::logic_error("When upstream_choke is set to false you should not supply a choke target group");
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Network/Node.cpp:40]: You should
```C++
#include <stdexcept>
```
to bring in a declaration of `std::logic_error`.

File: src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    std::vector<Branch> branches;
+    std::copy_if(this->m_branches.begin(), this->m_branches.end(), std::back_inserter(branches), [&node](const Branch& b) { return b.downtree_node() == node; });
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp:81]: > Could there be multiple Branches for which the .downtree_node() == node?

Then it is not a tree and `std::logic_error` will fire.

File: opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace Opm {
+namespace Network {
+
+class Node {
+public:
+    Node() = default;
+    Node(const std::string& name, std::optional<double> terminal_pressure, bool upstream_choke, bool add_gas_lift_gas, const std::string& choke_target_group);
+    Node(const std::string& name, std::optional<double> terminal_pressure, bool upstream_choke, bool add_gas_lift_gas);
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp]: OK - I have implemented it with setters.

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 8]
```diff
@@ -3116,6 +3126,70 @@ std::shared_ptr<const Python> Schedule::python() const
...
+            std::string target_group = name;
+            const auto& target_item = record.getItem<NP::CHOKE_GROUP>();
+            if (target_item.hasValue(0))
+                target_group = target_item.get<std::string>(0);
+
+            if (target_group != name) {
+                if (this->hasGroup(name, report_step)) {
+                    const auto& group = this->getGroup(name, report_step);
+                    if (group.numWells() > 0)
+                        throw std::invalid_argument("A manifold group must respond to it's own target");
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp:3162]: Nitpick:
```
...respond to ITS own...
```
Sorry.  didn't see this before.

File: opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (bska) [reply opm/parser/eclipse/EclipseState/Schedule/Network/Node.hpp]: > OK - I have implemented it with setters.

Thanks&mdash;much appreciated.

File: cmake/Modules/OpmDefaults.cmake [hunk 9]
```diff
@@ -5,6 +5,8 @@ is_compiler_gcc_compatible ()
 include(TestCXXAcceptsFlag)
 
 macro (opm_defaults opm)
+
+  message("Processing opm_defaults ${opm}")
```
Comment (bska) [inline cmake/Modules/OpmDefaults.cmake:9]: Is this leftover debugging code?

File: opm-common-prereqs.cmake [hunk 10]
```diff
@@ -15,6 +15,7 @@ set (opm-common_DEPS
 list(APPEND opm-common_DEPS
       # various runtime library enhancements
       "Boost 1.44.0 COMPONENTS system unit_test_framework REQUIRED"
+      "OpenMP QUIET"
```
Comment (bska) [inline opm-common-prereqs.cmake:18]: This doesn't seem to be related to the rest of code in this PR.

File: cmake/Modules/OpmDefaults.cmake (hunk 9 above)
Comment (joakim-hove) [reply cmake/Modules/OpmDefaults.cmake:9]: ?????? - this must be a rebase gone astray.

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 8 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: fixed

Comment (bska) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: > fixed

Thanks!

**Comment by bska** [comment]
Hm, we're getting linker errors of the form
```
/usr/bin/ld: CMakeFiles/test_ParallelRestart.dir/tests/test_ParallelRestart.cpp.o: in function `void Opm::Network::Branch::serializeOp<Opm::EclMpiSerializer>(Opm::EclMpiSerializer&)':
test_ParallelRestart.cpp:[...]: undefined reference to `unsigned long Opm::Mpi::packSize<double>(std::optional<double> const&, ompi_communicator_t*)'
```
Is there some down-stream PR we need to take into account?

**Comment by joakim-hove** [comment]
> Is there some down-stream PR we need to take into account?

yes - seems like it; I think this was green previously - but whatever:

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2577 please

**Comment by bska** [comment]
Green.  I'm merging this into master now.
